packages/react/src/components/Tree/Tree.tsx

```diff
@@ -0,0 +1,334 @@
+
+  /**
+   * A custom render function for the title slot.
+   *
+   * @param {React.ReactType} Component - The computed component for this slot.
+   * @param {object} props - The computed props for this slot.
+   * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+   */
+  renderItemTitle?: ShorthandRenderFunction
+}
+
+export interface TreeItemForRenderProps {
+  elementRef: React.RefObject<HTMLElement>
+  id: string
+  index: number
+  level: number
+  parent: ShorthandValue<TreeItemProps>
+  siblings: ShorthandCollection<TreeItemProps>
+}
+
+export interface TreeState {
+  activeItemIds: string[]
+  itemsForRender: { [s: string]: TreeItemForRenderProps }
+}
+
+class Tree extends AutoControlledComponent<WithAsProp<TreeProps>, TreeState> {
+  static create: Function
+
+  static displayName = 'Tree'
+
+  static className = 'ui-tree'
+
+  static slotClassNames: TreeSlotClassNames = {
+    item: `${Tree.className}__item`,
+  }
+
+  static propTypes = {
+    ...commonPropTypes.createCommon({
+      content: false,
+    }),
+    activeItemIds: customPropTypes.collectionShorthand,
+    defaultActiveItemIds: customPropTypes.collectionShorthand,
+    exclusive: PropTypes.bool,
+    items: customPropTypes.collectionShorthand,
+    renderItemTitle: PropTypes.func,
+    rtlAttributes: PropTypes.func,
+  }
+
+  static defaultProps = {
+    as: 'div',
+    accessibility: treeBehavior,
+  }
+
+  static autoControlledProps = ['activeItemIds']
+
+  static isSubtree(item: TreeItemProps | ShorthandValue<TreeItemProps>): boolean {
+    return !!item['items'] && item['items'].length > 0
+  }
+
+  static getItemsForRender = _.memoize((itemsFromProps: ShorthandCollection<TreeItemProps>) => {
+    const itemsForRenderGenerator = (
+      items = itemsFromProps,
+      level = 1,
+      parent?: ShorthandValue<TreeItemProps>,
+    ) => {
+      return _.reduce(
+        items,
+        (acc: Object, item: ShorthandValue<TreeItemProps>, index: number) => {
+          const id = item['id']
+          const isSubtree = Tree.isSubtree(item)
+
+          acc[id] = {
+            elementRef: React.createRef<HTMLElement>(),
+            level,
+            index,
+            parent,
+            siblings: items.filter(currentItem => currentItem !== item),
+          }
+
+          return {
+            ...acc,
+            ...(isSubtree ? itemsForRenderGenerator(item['items'], level + 1, item) : {}),
+          }
+        },
+        {},
+      )
+    }
+
+    return itemsForRenderGenerator(itemsFromProps)
+  })
+
+  static getAutoControlledStateFromProps(nextProps: TreeProps, prevState: TreeState) {
+    const itemsForRender = Tree.getItemsForRender(nextProps.items)
+
+    return {
+      itemsForRender,
+    }
+  }
+
+  getInitialAutoControlledState() {
+    return { activeItemIds: [] }
+  }
+
+  treeRef = React.createRef<HTMLElement>()
+
+  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
```

Comment:
```suggestion
  handleTreeItemOverrides = (predefinedProps: TreeItemProps): TreeItemProps => ({
```

It will also highlight an real issue:

![image](https://user-images.githubusercontent.com/14183168/63925418-6573f680-ca4a-11e9-8563-8d09babfbace.png)